src/Rules/Preg/RegexPatternRule.php

```diff
@@ -0,0 +1,119 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Preg;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\Constant\ConstantArrayType;
+use PHPStan\Type\Constant\ConstantStringType;
+
+class RegexPatternRule implements \PHPStan\Rules\Rule
+{
+
+	public function getNodeType(): string
+	{
+		return FuncCall::class;
+	}
+
+	/**
+	 * @param FuncCall $node
+	 * @param \PHPStan\Analyser\Scope $scope
+	 * @return string[]
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$patterns = $this->extractPatterns($node, $scope);
+
+		$errors = [];
+		foreach ($patterns as $pattern) {
+			if ($this->validatePattern($pattern)) {
+				continue;
+			}
+
+			$errors[] = sprintf('The regex pattern \'%s\' is not valid.', $pattern);
+		}
+
+		return $errors;
+	}
+
+	/**
+	 * @param FuncCall $functionCall
+	 * @param Scope $scope
+	 * @return string[]
+	 */
+	private function extractPatterns(FuncCall $functionCall, Scope $scope): array
+	{
+		if (!isset($functionCall->args[0])) {
+			return [];
+		}
+		$patternNode = $functionCall->args[0]->value;
+		$patternType = $scope->getType($patternNode);
+
+		if (!$functionCall->name instanceof Node\Name) {
+			return [];
+		}
+		$functionName = (string) $functionCall->name;
+
+		if (
+			$patternType instanceof ConstantStringType
```

Comment:
`$patternType` can also be a union of constant strings - use `TypeUtils::getConstantStrings` to get an array of those.

Reply:
I'm somewhat confused how `getConstantStrings` works/should work. For example:

```php
$string = random_int(0, 1) === 0
	? '~ok~'
	: 123;
```

This will result in the type `123|'~ok~'`. Running `getConstantStrings` results in an empty array. This is implemented [here](https://github.com/phpstan/phpstan/blob/1401d9a728be947f8dd82055f72ca8073575fa44/src/Type/TypeUtils.php#L112-L123). Shouldn't the result be `['123']`?

Furthermore, if you do something like:

```php
$string = random_int(0, 1) === 0
	? '~ok~'
	: $unknownString;
```

This will result in just `string` instead of `string|'~ok~'`, even before passing it to `getConstantStrings`.

Is this the expected behavior?

Reply:
You're not getting `123|'~ok~'` because that's how the method `getConstantStrings` is designed. if the union consists only of constant strings then you get them, otherwise you get an empty array. It's an assumption that allows this method to be used in places where it's used. I think that by using it you will make your code more general. `Having 123|'foo'` will result in an error on level 6 when calling function that accepts only string.

Reply:
You're not getting `string|'~ok~'` because that's a combination that cannot exist with the current typesystem. string is a supertype of `~ok~`, so the constant string is eliminated during normalization in TypeCombinator. Same as `mixed` cannot be combined with anything because it's a top type.

Reply:
@ondrejmirtes I see. `string|'~ok~'` would be nice because we could analyze `'~ok~'`. But I'm thinking 90% of the time the regex will be passed directly or come from a constant. So this should work just fine.